lib/web_ui/test/composition_test.dart

```diff
@@ -0,0 +1,165 @@
+            MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = fakeComposingText;
+        mockWithCompositionAwareMixin.addCompositionEventHandlers(inputElement);
+
+        inputElement.dispatchEvent(html.Event(CompositionAwareMixin.kCompositionEnd));
+
+        expect(mockWithCompositionAwareMixin.composingText, null);
+      });
+    });
+
+    group('composition start', () {
+      test('should reset composing text on handle composition start', () {
+        final MockWithCompositionAwareMixin mockWithCompositionAwareMixin =
+            MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = fakeComposingText;
+        mockWithCompositionAwareMixin.addCompositionEventHandlers(inputElement);
+
+        inputElement.dispatchEvent(html.Event(CompositionAwareMixin.kCompositionStart));
+
+        expect(mockWithCompositionAwareMixin.composingText, null);
+      });
+    });
+
+    group('composition update', () {
+      test('should set composing text to event composing text', () {
+        const String fakeEventText = 'IAmComposingThis';
+        final MockWithCompositionAwareMixin mockWithCompositionAwareMixin =
+            MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = fakeComposingText;
+        mockWithCompositionAwareMixin.addCompositionEventHandlers(inputElement);
+
+        inputElement.dispatchEvent(
+            html.CompositionEvent(CompositionAwareMixin.kCompositionUpdate, data: fakeEventText));
+
+        expect(mockWithCompositionAwareMixin.composingText, fakeEventText);
+      });
+    });
+
+    group('determine composition state', () {
+      test('should return new composition state if valid new composition', () {
+        const int baseOffset = 100;
+        const String composingText = 'composeMe';
+
+        final EditingState editingState = EditingState(
+          extentOffset: baseOffset,
+          text: 'testing',
+          baseOffset: baseOffset,
+        );
+
+        final MockWithCompositionAwareMixin mockWithCompositionAwareMixin =
+            MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = composingText;
+
+        const int expectedComposingBase = baseOffset - composingText.length;
+
+        expect(
+            mockWithCompositionAwareMixin.determineCompositionState(editingState),
+            editingState.copyWith(
+                composingBaseOffset: expectedComposingBase,
+                composingExtentOffset: expectedComposingBase + composingText.length));
+      });
+    });
+  });
+
+  group('composing range', () {
+    test('should be [0, compostionStrLength] on new composition', () {
+      const String composingText = 'hi';
+
+      inputElement.dispatchEvent(
+          html.CompositionEvent(CompositionAwareMixin.kCompositionUpdate, data: composingText));
+
+      //set the selection text
```

Comment:
Space after the `//`, capital letter at the start and period at the end, here and a few other comments below.